<a href="https://colab.research.google.com/github/ArredondoF99/VIOLENCIA-FAMILIAR-EN-TIEMPOS-DE-COVID-19/blob/main/Proyecto_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Casos de violencia familiar y su relación con el aislamiento por COVID-19**

###Integrantes

* Francisco Ricardo Arredondo Almuina - frankk_arredondo@hotmail.com
* Jesus Emanuel Serrano Molina - emanuel.serm@gmail.com
* Carlos Eduardo Vidal Villeda - charlie.lalo@hotmail.com

#  **IDENTIFICACIÓN DEL PROBLEMA**

## Problemáticas

1. Violencia familiar en tiempos de COVID – 19

2. Violencia de género en tiempos de COVID – 19

## Investigación 
- Díaz Olavarrieta, C., & Phillips, V. (s.f.). COVID-19 y Violencia Basada en Género/Violencia Intrafamiliar. Recuperado el 24 de Febrero de 2021, de ACADEMIA NACIONAL DE MEDICINA DE MÉXICO: https://www.anmm.org.mx/descargas/COVID-19-y-Violencia-Basada-en-Genero-Violencia-Intrafamiliar.pdf

- EQUIS Justicia para las Mujeres, l. R. (s.f.). LAS DOS PANDEMIAS: VIOLENCIA CONTRA LAS MUJERES EN MÉXICO EN EL CONTEXTO DEL COVID-19. Recuperado el 24 de Febrero de 2021, de EQUIS Justicia para las Mujeres: https://equis.org.mx/wp-content/uploads/2020/08/informe-dospandemiasmexico.pdf

- Fenández Nicto, B. (3 de Mayo de 2020). Violencia Doméstica en México durante el confinamiento por la pandemia COVID-19. (I. Yánez Soria, & M. Antonia Bravo, Editores) Recuperado el 24 de Febrero de 2021, de DATA-POP ALLIANCE: https://datapopalliance.org/violencia-domestica-en-mexico-durante-el-confinamiento-por-la-pandemia-covid-19/

- Gómez Macfarland, C. A., & Sánchez Ramírez, M. C. (Junio de 2020). Violencia familiar en tiempos de Covid. Recuperado el 24 de Febrero de 2021, de Instituto Belisario Domínguez: http://bibliodigitalibd.senado.gob.mx/bitstream/handle/123456789/4891/ML_187.pdf?sequence=1&isAllowed=y

- Llamas Caballero, R. (s.f.). Violencia doméstica y de género en tiempos de COVID-19. Recuperado el 24 de Febrero de 2021, de Anáhuac México: https://www.anahuac.mx/mexico/noticias/Violencia-domestica-y-de-genero-en-tiempos-de-COVID-19

- Monroy, J. (16 de Abril de 2020). Segob: violencia intrafamiliar aumentó 120% desde la emergencia del Covid-19. Recuperado el 24 de Febrero de 2021, de EL ECONOMISTA: https://www.eleconomista.com.mx/politica/Segob-violencia-intrafamiliar-aumento-120-desde-la-emergencia-del-Covid-19-20200416-0111.html

- Núñez Trejo, A. (7 de Junio de 2020). Crece la violencia intrafamiliar en México, por el COVID-19. Recuperado el 24 de Febrero de 2021, de FORO JURÍDICO: https://forojuridico.mx/crece-la-violencia-intrafamiliar-en-mexico-por-el-covid-19/

- Rodríguez Pedraza, Y. (2020). La feminización de la pandemia COVID19. Recuperado el 24 de Febrero de 2021, de Redalyc: https://www.redalyc.org/articulo.oa?id=29063559023

## Medidas ante las problemáticas 



*   Órdenes de protección.
*   Si identifica que usted experimenta una de las características del ciclo de la violencia, acuda al Agente del Ministerio Público y busque ayuda.
*  Resguarde sus documentos, así como los de sus hijos, en una casa de algún familiar.
*  Si su pareja o el agresor se pone violento, retírese inmediatamente.
*  Acuda a buscar ayuda a las instituciones especializadas (UAVI, Unidad de Atención de Violencia Intrafamiliar, IJM Instituto Jalisciense de las Mujeres, DIF Municipales) para buscar refugio temporal
*  Llame al 066 para pedir ayuda la policía.

#### *Fuentes para medidas ante las problemáticas*

- Hernández, R. (s.f.). MEDIDAS PREVENTIVAS PARA EVITAR LA VIOLENCIA INTRAFAMILIAR. Recuperado el 25 de Febrero de 2021, de FISCALIA GENERAL DEL ESTADO. GOBIERNO DEL ESTADO DE JALISCO: https://fge.jalisco.gob.mx/prevencion-social/medidas-preventivas-para-evitar-la-violencia-intrafamiliar



# **PLANTEAMIENTO DE PREGUNTAS CLAVE**

### 1. ¿Cuáles son los estados con mayor violencia familiar en México?

### 2. ¿Cuáles son los estados con menos incidencia de violencia familiar?

### 3. ¿Cuál es el mes con mayor incidencia de violencia familiar?

### 4. ¿Cuál es el año con mayor incidencia de violencia familiar?

### 5. ¿Es cierto que en el 2020 el crecimiento en incidencia de violencia familiar aumento debido a la pandemia?

### 6. ¿Este fenómeno se repitió en otros países o solo en México?

# **COLECCIÓN DE DATOS**
La colección de datos es una parte fundamental dentro de un proyecto de ciencia de datos, siendo que es importante tener una fuente confiable y segura de los datos que van a ser sometidos a un análisis. Existen distintas fuentes para extraer bases de datos, entre las más destacadas se encuentran sitios como Kaggle, APIs públicas o privadas, Instituciones gubernamentales o privadas, entre otras.

Siendo que nuestro proyecto esta dirigido a analizar la "Violencia familiar en tiempos de COVID-19" y la "Violencia de género en tiempos de COVID-19", es fundamental obtener nuestros datos de una fuente confiable y que realmente nos ayude a poder analizar los delitos cometidos en México.

Para la obtención de los datos en cuanto a la República Mexicana fue usado el website del Gobierno de México, dado que para poder obtener los delitos en cada uno de los Estados y Municipios del país, deben de ser obtenido mediante una entidad gubernamental por temas de confidencialidad y resguardo de datos, por lo que en la siguiente dirección URL se encuentran los datos abiertos de incidencia delictiva en los últimos 5 años:

https://www.gob.mx/sesnsp/acciones-y-programas/datos-abiertos-de-incidencia-delictiva

Cabe destacar que de dicha dirección se obtuvieron dos archivos .CSV con los distintos delitos cometidos en los diferentes Estados de la República Mexicana, asimismo como en cada Municipio de los mismos, lo cual posteriormente nos brindará un análisis más completo de acuerdo con el fin del proyecto.

Para leer lor achivos mencionados en Python, es posible realizarlo mediante dos maneras, importandolo desde los archivos guardados de la computadora, o previamente haberlo guardado en Google Drive e importarlo directamente a Google Colaboratory. Ambas maneras consiguen obtener los archivos .CSV que servirán para realizar un Análisis Exploratorio, Limpieza y Transformación de nuestros datos.

###**Import from Files**

Primeramente se importaron las librerías necesarias.

In [8]:
import pandas as pd

Después, tras ya tener los archivos correspondientes en nuestra computadora, correremos el siguiente código para poder elegir estos archivos y que Google colaboratory pueda leerlos y almacenarlos en los archivos de la Notebook.

In [ ]:
from google.colab import files
files.upload()

Tras almacenar los datos dentro de los archivos del Notebook, copiaremos la ruta de los mismos para poder leerlos mediante pandas. Nombrando al dataframe con los datos de delincuencia de los Estados y Municipios de México como "df_estados" y "df_municipios", respectivamente.

Cabe destacar que para que puedan ser leidos los documentos estos deben de tener un encoding de "latin-1" para que python pueda leerlo en base al idioma de los datos. 

In [10]:
df_estados = pd.read_csv("/content/Estatal-Delitos-2015-2021_ene2021.csv", encoding="latin-1")
df_municipios = pd.read_csv("/content/Municipal-Delitos-2015-2021_ene2021.csv", encoding="latin-1")

Finalmente para ver que nuestro dataframes fueron almacenados correctamente y explorar un poco su información se usó el comando "head"

In [11]:
df_estados.head(2)

Año  Clave_Ent         Entidad  ... Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...       2       2.0       1.0
1  2015          1  Aguascalientes  ...       0       0.0       1.0

[2 rows x 19 columns]

In [12]:
df_municipios.head(2)

Año  Clave_Ent         Entidad  ...  Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...      1.0       0.0       1.0
1  2015          1  Aguascalientes  ...      0.0       0.0       0.0

[2 rows x 21 columns]

###**Import from Drive**

Primeramente se importaron las librerías necesarias, tal cual el caso anterior.

In [13]:
import pandas as pd

A diferencia de importar los archivos desde nuestra computadora, aquí son almacenados previamente desde Google Drive, y mediante el siguiente código podemos traerlos a los archivos de nuestra Notebook. 

Cabe destacar que la dirección de la cual son obtenidos los archivos desde Google Drive varía de acuerdo a donde esté almacenado, por lo que en nuestro caso fue el siguiente.

In [14]:
from google.colab import drive
drive.mount('/content/drive/')
!ls "/content/drive/MyDrive/Data_Proy"

Mounted at /content/drive/
ls: cannot access '/content/drive/MyDrive/Data_Proy': No such file or directory


Una vez teniendo dichos archivos ya en el Notebook, se realiza la misma secuencia para leer los archivos .CSV mediante pandas, nombrándolos como "df_estados" y "df_municipios".

In [18]:
df_estados = pd.read_csv("/content/drive/MyDrive/Proyecto Python/Estatal-Delitos-2015-2021_ene2021.csv", encoding="latin-1")
df_municipios = pd.read_csv("/content/drive/MyDrive/Proyecto Python/Municipal-Delitos-2015-2021_ene2021.csv", encoding="latin-1")

Finalmente para ver que nuestro dataframes fueron almacenados correctamente y explorar un poco su información se usó el comando "head"

In [19]:
df_estados.head(2)

Año  Clave_Ent         Entidad  ... Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...       2       2.0       1.0
1  2015          1  Aguascalientes  ...       0       0.0       1.0

[2 rows x 19 columns]

In [20]:
df_municipios.head(2)

Año  Clave_Ent         Entidad  ...  Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...      1.0       0.0       1.0
1  2015          1  Aguascalientes  ...      0.0       0.0       0.0

[2 rows x 21 columns]

# **ANÁLISIS EXPLORATORIO DE DATOS**

Una vez que hemos importado nuestros archivos vamos a empezar a analizar nuestro conjunto de datos usando pandas. Nos ayudamos con algunas preguntas que nos indicarán si nuestra información recolectada nos será de ayuda o no.


##  **1. ¿El conjunto de datos que tengo realmente me sirve para responder algunas de las preguntas que me planteé?**

In [21]:
#previamente se cargaron nuestros archivos csv y se asignaron a las variables df_estados y df_municipios
df_estados = pd.DataFrame(df_estados)
df_estados.head(2)

Año  Clave_Ent         Entidad  ... Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...       2       2.0       1.0
1  2015          1  Aguascalientes  ...       0       0.0       1.0

[2 rows x 19 columns]

In [22]:
df_municipios = pd.DataFrame(df_municipios)
df_municipios.head(2)

Año  Clave_Ent         Entidad  ...  Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...      1.0       0.0       1.0
1  2015          1  Aguascalientes  ...      0.0       0.0       0.0

[2 rows x 21 columns]

In [23]:
df_municipios.tail(2)

Año  Clave_Ent    Entidad  ...  Octubre Noviembre Diciembre
1588774  2021         32  Zacatecas  ...      NaN       NaN       NaN
1588775  2021         32  Zacatecas  ...      NaN       NaN       NaN

[2 rows x 21 columns]

_*Con un análisis rápido de nuestros data sets encotramos datos que incluyen información relevante a nuestra investigación, como lo son: tipos de delitos, estado de la República mexicana en donde ocurrieron y fecha en la que ocurrieron.
Por lo tanto podemos ser positivos de que nuestras preguntas serán respondidas*_


## **2. ¿Qué tamaño tiene mi conjunto de datos? ¿Serán datos suficientes?**

In [24]:
df_estados.shape

(21952, 19)

In [25]:
df_municipios.shape

(1588776, 21)

*Vemos que nuestro data frame **`df_estados`** tiene 19 columnas y 21952 filas,, mientras que el data frame **`df_municipios`** cuenta con 21 columnas y 1588776 filas, esta diferencia de filas es posible que se deba a que nuestro segundo data frame agrega la información en donde los actos de violencia fueron registrados (columna de municipios).
Debido a esto y a que ambos data sets fueron obtenidos del mismo lugar, en adelante solo se considerará **`df_municipios`** ya que la información está más completa agragando el campo de los municipios, mientras que **`df_estados`** es una recopilación de estos y estaríamos cometiendo un error de replica de datos*

## **3. ¿Qué columnas tengo y qué información tengo en cada una de esas columnas?**

In [26]:
df_municipios.columns

Index(['Año', 'Clave_Ent', 'Entidad', 'Cve. Municipio', 'Municipio',
       'Bien jurídico afectado', 'Tipo de delito', 'Subtipo de delito',
       'Modalidad', 'Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio',
       'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre'],
      dtype='object')

- En la columna **`Año`** tenemos registros que empiezan en el 2015 y llegan hasta el 2021.
- En **`Entidad`** yacen los nombres de los estados donde ocurren los hechos mientras que **`Caleve_Ent`** es la calve federativa de este estado.
- Así en **`Municipio`** se agregan el municipio del estado correspondiente y **`Cve. Municipio`** es su clave única.
- **`Bien jurídico afectado`** , **`Tipo de delito`**, **`Subtipo de delito`** y **`Modalidad`** son columnas que describen los delitos registrados.
- Las columnas de **`Enero`**, **`Febrero`**, **`Marzo`**, **`Abril`**, **`Mayo`**, **`Junio`**, **`Julio`**, **`Agosto`**, **`Septiembre`**, **`Octubre`**, **`Noviembre`**, **`Diciembre`** son las sumas acumuladas de los registros por tipos de delito.


## **4. Los nombres que tienen mis columnas, ¿son el nombre más apropiado?** 

- *Si, son descriptivas, simples y van de acuerdo a a la información que queremos obtener*

## **5. ¿Qué tipos de datos tengo en cada columna? ¿Parecen ser el tipo correcto de datos? ¿O es un tipo de datos "incorrecto"?** 

In [27]:
df_municipios.dtypes

Año                         int64
Clave_Ent                   int64
Entidad                    object
Cve. Municipio              int64
Municipio                  object
Bien jurídico afectado     object
Tipo de delito             object
Subtipo de delito          object
Modalidad                  object
Enero                       int64
Febrero                   float64
Marzo                     float64
Abril                     float64
Mayo                      float64
Junio                     float64
Julio                     float64
Agosto                    float64
Septiembre                float64
Octubre                   float64
Noviembre                 float64
Diciembre                 float64
dtype: object

- *Las columnas tienen el tipo de dato correcto, especialmente nos interesa que las columnas de fecha como `Año` y los casos acumulados sean datos de tipo númerico.*

## **6. Si selecciono algunas filas al azar y las observo, ¿estoy obteniendo los datos que debería? ¿o hay datos que parecen estar "sucios" o "incorrectos"?** 

In [28]:
df_municipios.loc[[0,182,1112,45932,1588775], ['Año', 'Entidad', 'Municipio', 'Tipo de delito','Modalidad', 'Enero' ]]

Año  ... Enero
0        2015  ...     2
182      2015  ...     0
1112     2015  ...     4
45932    2015  ...     0
1588775  2021  ...     0

[5 rows x 6 columns]

- *Hasta el momento todo parece ser correcto, los datos no muestran estar sucios*

# **LIMPIEZA DE DATOS**



In [29]:
#Este el el dataframe de los datos organizados por municipio
df_municipios

Año  Clave_Ent         Entidad  ...  Octubre Noviembre Diciembre
0        2015          1  Aguascalientes  ...      1.0       0.0       1.0
1        2015          1  Aguascalientes  ...      0.0       0.0       0.0
2        2015          1  Aguascalientes  ...      0.0       0.0       0.0
3        2015          1  Aguascalientes  ...      0.0       0.0       0.0
4        2015          1  Aguascalientes  ...      0.0       0.0       0.0
...       ...        ...             ...  ...      ...       ...       ...
1588771  2021         32       Zacatecas  ...      NaN       NaN       NaN
1588772  2021         32       Zacatecas  ...      NaN       NaN       NaN
1588773  2021         32       Zacatecas  ...      NaN       NaN       NaN
1588774  2021         32       Zacatecas  ...      NaN       NaN       NaN
1588775  2021         32       Zacatecas  ...      NaN       NaN       NaN

[1588776 rows x 21 columns]

In [30]:
#En este bloque se analiza si existe prescencias de nans en nuetsro dataframe
df_municipios.isna().sum(axis=0)

Año                            0
Clave_Ent                      0
Entidad                        0
Cve. Municipio                 0
Municipio                      0
Bien jurídico afectado         0
Tipo de delito                 0
Subtipo de delito              0
Modalidad                      0
Enero                          0
Febrero                   242158
Marzo                     242158
Abril                     242158
Mayo                      242158
Junio                     242158
Julio                     242158
Agosto                    242158
Septiembre                242158
Octubre                   242158
Noviembre                 242158
Diciembre                 242158
dtype: int64

In [31]:
#En este bloque se elimina los nans
df_municipios_limpio= df_municipios.dropna(axis=0, how='any')

In [32]:
#Aqui se corrobora si el codigo anterior se ejecuto de manera correcta
df_municipios_limpio.isna().sum(axis=0)

Año                       0
Clave_Ent                 0
Entidad                   0
Cve. Municipio            0
Municipio                 0
Bien jurídico afectado    0
Tipo de delito            0
Subtipo de delito         0
Modalidad                 0
Enero                     0
Febrero                   0
Marzo                     0
Abril                     0
Mayo                      0
Junio                     0
Julio                     0
Agosto                    0
Septiembre                0
Octubre                   0
Noviembre                 0
Diciembre                 0
dtype: int64

In [33]:
df_municipios_limpio

Año  Clave_Ent         Entidad  ...  Octubre Noviembre Diciembre
0        2015          1  Aguascalientes  ...      1.0       0.0       1.0
1        2015          1  Aguascalientes  ...      0.0       0.0       0.0
2        2015          1  Aguascalientes  ...      0.0       0.0       0.0
3        2015          1  Aguascalientes  ...      0.0       0.0       0.0
4        2015          1  Aguascalientes  ...      0.0       0.0       0.0
...       ...        ...             ...  ...      ...       ...       ...
1346613  2020         32       Zacatecas  ...      0.0       0.0       0.0
1346614  2020         32       Zacatecas  ...      0.0       0.0       0.0
1346615  2020         32       Zacatecas  ...      0.0       0.0       0.0
1346616  2020         32       Zacatecas  ...      0.0       0.0       0.0
1346617  2020         32       Zacatecas  ...      1.0       0.0       1.0

[1346618 rows x 21 columns]

In [34]:
#En esta parte scaamos la suma por mes de todos los años
mes=['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre','Octubre','Noviembre','Diciembre']
for i in range(0,11):
  print('La suma del mes de ', mes[i], ' = ',  df_municipios_limpio[mes[i]].sum())

La suma del mes de  Enero  =  894267
La suma del mes de  Febrero  =  896080.0
La suma del mes de  Marzo  =  970568.0
La suma del mes de  Abril  =  902224.0
La suma del mes de  Mayo  =  951787.0
La suma del mes de  Junio  =  949497.0
La suma del mes de  Julio  =  948792.0
La suma del mes de  Agosto  =  973212.0
La suma del mes de  Septiembre  =  950207.0
La suma del mes de  Octubre  =  1001677.0
La suma del mes de  Noviembre  =  928524.0


In [35]:
#En esta parte sacamos el minimo por mes de todos los años
for i in range(0,11):
  print('El min del mes de ', mes[i], ' =',  df_municipios_limpio[mes[i]].min())

El min del mes de  Enero  = 0
El min del mes de  Febrero  = 0.0
El min del mes de  Marzo  = 0.0
El min del mes de  Abril  = 0.0
El min del mes de  Mayo  = -1.0
El min del mes de  Junio  = 0.0
El min del mes de  Julio  = 0.0
El min del mes de  Agosto  = 0.0
El min del mes de  Septiembre  = -1.0
El min del mes de  Octubre  = 0.0
El min del mes de  Noviembre  = 0.0


In [36]:
#Nos percatamos que habia un error en nuestro dataframe donde habia un -1 cuando
#suponemos que tendria que ser 1 por ende tenemos que limpiar los datos nuevamente

In [37]:
df_municipios_limpio['Mayo']=df_municipios_limpio['Mayo'].replace([-1],1)
df_municipios_limpio['Septiembre']=df_municipios_limpio['Septiembre'].replace([-1],1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [38]:
#En esta parte sacamos el minimo por mes de todos los años
for i in range(0,11):
  print('El min del mes de ', mes[i], ' =',  df_municipios_limpio[mes[i]].min())

El min del mes de  Enero  = 0
El min del mes de  Febrero  = 0.0
El min del mes de  Marzo  = 0.0
El min del mes de  Abril  = 0.0
El min del mes de  Mayo  = 0.0
El min del mes de  Junio  = 0.0
El min del mes de  Julio  = 0.0
El min del mes de  Agosto  = 0.0
El min del mes de  Septiembre  = 0.0
El min del mes de  Octubre  = 0.0
El min del mes de  Noviembre  = 0.0


In [39]:
#En esta parte sacamos el maximo por mes de todos los años
for i in range(0,11):
  print('El max del mes de ', mes[i], ' =',  df_municipios_limpio[mes[i]].max())

El max del mes de  Enero  = 1146
El max del mes de  Febrero  = 1212.0
El max del mes de  Marzo  = 1471.0
El max del mes de  Abril  = 1445.0
El max del mes de  Mayo  = 1444.0
El max del mes de  Junio  = 958.0
El max del mes de  Julio  = 1219.0
El max del mes de  Agosto  = 1140.0
El max del mes de  Septiembre  = 1099.0
El max del mes de  Octubre  = 1023.0
El max del mes de  Noviembre  = 1156.0


# **TRANSFORMACIÓN DE DATOS**

Ha llegado el momento de aplicar técnicas comunes de procesamiento de datos a nuestro conjunto de datos. Para preparar la información para un posterior análisis.

Con base en los previos pasos de nuestro proyecto, hasta el momento tenemos un dataframe llamado `df_estados_limpio` en el cual tenemos la información de cada uno de los Estados, asímismo como de los Municipios, de todos los delitos que se han cometido desde 2015 hasta la fecha por mes de acuerdo al tipo de delito. 

Tras ya tener el dataframe limpio podemos pasar a la filtración, transformación y reordenamiento de nuestros datos. 

Primeramente necesitamos ver cuales son los tipos de dato para cada columna y si es que estos son correctos.

In [40]:
df_municipios_limpio.dtypes

Año                         int64
Clave_Ent                   int64
Entidad                    object
Cve. Municipio              int64
Municipio                  object
Bien jurídico afectado     object
Tipo de delito             object
Subtipo de delito          object
Modalidad                  object
Enero                       int64
Febrero                   float64
Marzo                     float64
Abril                     float64
Mayo                      float64
Junio                     float64
Julio                     float64
Agosto                    float64
Septiembre                float64
Octubre                   float64
Noviembre                 float64
Diciembre                 float64
dtype: object

Como podemos observar tras usar el comando *dtypes* a nuestro dataframe, todas las columnas tienen un tipo de dato correcto, únicamente las columnas de los meses de Febrero a Diciembre tienen un tipo de dato flotante que hace referencia a los casos delictivos que hubo en esos meses. 

A primer instancia podríamos considerar correctos siendo tipo *float64* dado que son números las cantidades de casos en cada mes, pero lo más correcto es considerar estas cantidades como números enteros, por lo cual deben de estar como tipo de dato *int64* como está la columna del mes de Enero.

Para esto crearemos un diccionario de conversión en el cual colocaremos cada uno de los meses con el nuevo tipo de dato que queremos obtener, siendo *int64*.

In [41]:
diccionario_de_conversion = {
    "Febrero": "int64",
    "Marzo": "int64",
    "Abril": "int64",
    "Mayo": "int64",
    "Junio": "int64",
    "Julio": "int64",
    "Agosto": "int64",
    "Septiembre": "int64",
    "Octubre": "int64",
    "Noviembre": "int64",
    "Diciembre": "int64",
}

Posteriormente, usaremos el comando *astype* para que de nuestro dataframe de `df_municipios_limpio` seleccione las columnas que se encuentran en el `diccionario_de_conversion` y las convierta en el nuevo tipo de dato. Renombraremos a este nuevo dataframe con los tipo de datos correctos, como `df_municipios` (anteriormente asignado) por temas de simplicidad en el nombre de las variables. 

Una vez hecho esto, podremos darnos cuenta que al volver a correr el código *dtypes*, nuestras columnas de los meses tienen el tipo de dato correctamente.

In [42]:
df_municipios = df_municipios_limpio.astype(diccionario_de_conversion)
df_municipios.dtypes

Año                        int64
Clave_Ent                  int64
Entidad                   object
Cve. Municipio             int64
Municipio                 object
Bien jurídico afectado    object
Tipo de delito            object
Subtipo de delito         object
Modalidad                 object
Enero                      int64
Febrero                    int64
Marzo                      int64
Abril                      int64
Mayo                       int64
Junio                      int64
Julio                      int64
Agosto                     int64
Septiembre                 int64
Octubre                    int64
Noviembre                  int64
Diciembre                  int64
dtype: object

In [43]:
df_municipios.head(2)

Año  Clave_Ent         Entidad  ...  Octubre Noviembre Diciembre
0  2015          1  Aguascalientes  ...        1         0         1
1  2015          1  Aguascalientes  ...        0         0         0

[2 rows x 21 columns]

Después, dentro de la estética de nuestros datos, tenemos que algunos nombres de columnas no se encuentran dentro del "formato recomendado" para el código por Python, en el cual se usan minúsculas y palabras separadas por guiones bajos, además de que la columna "Cve. Municipio" puede llegar a malinterpretarse por el acrónimo que hace referencia a la palabra "Clave". Dicho esto renombraremos algunas de las columnas usando un diccionario al cual llamaremos *diccionario_de_renombramiento*.

In [44]:
diccionario_de_renombramiento = {
    "Año": "año",
    "Clave_Ent": "clave_entidad",
    "Entidad": "entidad",
    "Cve. Municipio": "clave_municipio",
    "Municipio": "municipio",
    "Bien jurídico afectado": "bien_juridico_afectado",
    "Tipo de delito": "tipo_delito",
    "Subtipo de delito": "subtipo_delito",
    "Modalidad": "modalidad",
    "Enero": "enero",
    "Febrero": "febrero",
    "Marzo": "marzo",
    "Abril": "abril",
    "Mayo": "mayo",
    "Junio": "junio",
    "Julio": "julio",
    "Agosto": "agosto",
    "Septiembre": "septiembre",
    "Octubre": "octubre",
    "Noviembre": "noviembre",
    "Diciembre": "diciembre"
}

Con el diccionario de renombramiento nosotros asignamos mediante una llave-valor, el nuevo nombre de las columnas asignadas a cada nombre previo, por lo que usando el comando *rename*, se asigna el renombramiento de las columnas a nuestro dataframe *df_municipios* y al sobrescribirlo en el mismo dataframe veremos que las columnas han sido cambiadas a nuestro formato.

In [45]:
df_municipios = df_municipios.rename(columns=(diccionario_de_renombramiento))
df_municipios.head(2)

año  clave_entidad         entidad  ...  octubre noviembre diciembre
0  2015              1  Aguascalientes  ...        1         0         1
1  2015              1  Aguascalientes  ...        0         0         0

[2 rows x 21 columns]

Ahora si, tras haber transformado nuestros datos y renombrado las columnas, podemos pasar a la filtración y reordenamiento de los mismos. Actualmente el dataframe que tenemos, tiene todos los delitos cometidos desde 2015 al mes de Enero del 2021, sin embargo para el análisis de nuestros datos únicamente necesitamos los delitos relacionados a la "violencia familiar" y "violencia de género" como ha estipulado en el objetivo del proyecto. 

Para esto hay que conocer cuales son los delitos que se encuentran en la nueva columna "tipo_delito", usando el comando *unique*.

In [46]:
df_municipios["tipo_delito"].unique()

array(['Homicidio', 'Lesiones', 'Feminicidio', 'Aborto',
       'Otros delitos que atentan contra la vida y la integridad corporal',
       'Secuestro', 'Tráfico de menores', 'Rapto',
       'Otros delitos que atentan contra la libertad personal',
       'Abuso sexual', 'Acoso sexual', 'Hostigamiento sexual',
       'Violación simple', 'Violación equiparada', 'Incesto',
       'Otros delitos que atentan contra la libertad y la seguridad sexual',
       'Robo', 'Fraude', 'Abuso de confianza', 'Extorsión',
       'Daño a la propiedad', 'Despojo',
       'Otros delitos contra el patrimonio', 'Violencia familiar',
       'Violencia de género en todas sus modalidades distinta a la violencia familiar',
       'Incumplimiento de obligaciones de asistencia familiar',
       'Otros delitos contra la familia', 'Corrupción de menores',
       'Trata de personas', 'Otros delitos contra la sociedad',
       'Narcomenudeo', 'Amenazas', 'Allanamiento de morada',
       'Evasión de presos', 'Falsedad'

Como podemos observar, dentro de los delitos que se encuentran en el dataframe, se encuentran "Violencia familiar" y "Violencia de género en todas sus modalidades distinta a la violencia familiar", que son justo los delitos que queremos someter a un futuro análisis, por lo que usaremos un filtro para dividir nuestro dataframe en dos `df_violencia_familiar` y  `df_violencia_genero`.

In [47]:
df_violencia_familiar = df_municipios[df_municipios["tipo_delito"] == "Violencia familiar"]
df_violencia_familiar.head(10)

año  clave_entidad         entidad  ...  octubre noviembre diciembre
81   2015              1  Aguascalientes  ...       53        40        39
179  2015              1  Aguascalientes  ...        1         1         0
277  2015              1  Aguascalientes  ...        1         0         1
375  2015              1  Aguascalientes  ...        0         0         0
473  2015              1  Aguascalientes  ...        1         1         0
571  2015              1  Aguascalientes  ...        0         0         0
669  2015              1  Aguascalientes  ...        0         1         0
767  2015              1  Aguascalientes  ...        0         0         0
865  2015              1  Aguascalientes  ...        0         0         0
963  2015              1  Aguascalientes  ...        1         0         0

[10 rows x 21 columns]

In [48]:
df_violencia_genero = df_municipios[df_municipios["tipo_delito"] == "Violencia de género en todas sus modalidades distinta a la violencia familiar"]
df_violencia_genero.head(10)

año  clave_entidad         entidad  ...  octubre noviembre diciembre
82   2015              1  Aguascalientes  ...        1         0         0
180  2015              1  Aguascalientes  ...        0         0         0
278  2015              1  Aguascalientes  ...        0         0         0
376  2015              1  Aguascalientes  ...        0         0         0
474  2015              1  Aguascalientes  ...        0         0         0
572  2015              1  Aguascalientes  ...        0         0         0
670  2015              1  Aguascalientes  ...        0         0         0
768  2015              1  Aguascalientes  ...        0         0         0
866  2015              1  Aguascalientes  ...        0         0         0
964  2015              1  Aguascalientes  ...        0         0         0

[10 rows x 21 columns]

Sin embargo, si nos cuenta nuestro dataframe conserva el índice que tenía anteriormente, por lo cual usaremos el comando *reset_index* para tener uno nuevo secuencial y ordenado de acuerdo a los filtros hechos, de igual manera se guardará sobrescribiendo ambos dataframes "df_violencia_familiar" y "df_violencia_genero".

In [49]:
df_violencia_familiar = df_violencia_familiar.reset_index(drop=True)
df_violencia_familiar.head(3)

año  clave_entidad         entidad  ...  octubre noviembre diciembre
0  2015              1  Aguascalientes  ...       53        40        39
1  2015              1  Aguascalientes  ...        1         1         0
2  2015              1  Aguascalientes  ...        1         0         1

[3 rows x 21 columns]

In [50]:
df_violencia_familiar.to_csv("violencia_familiar.csv")

In [51]:
df_violencia_genero = df_violencia_genero.reset_index(drop=True)
df_violencia_genero.head(3)

año  clave_entidad         entidad  ...  octubre noviembre diciembre
0  2015              1  Aguascalientes  ...        1         0         0
1  2015              1  Aguascalientes  ...        0         0         0
2  2015              1  Aguascalientes  ...        0         0         0

[3 rows x 21 columns]

In [52]:
df_violencia_genero.to_csv("violencia_genero.csv")

Llegado a este punto, tenemos dos dataframes que serán la base para un futúro análisis tras conocer diferentes herramientas en futuro módulos para Python, tales como representación gráfica, series de tiempo, estadística, entre otros.

No obstante, para responder algunas pregunas planteadas en el inicio del proyecto, realizaremos un pequeño análisis con las herramientas que tenemos actualmente, en donde seleccionaremos un estado para filtrar los datos y conocer los máximos y mínimos tanto mensuales como anuales, dicho estado será Ciudad de México y será presentado a continuación.

Sin embargo, solo nos enfocaremos en el dataframe que hace referencia a la "Violencia familiar" siendo que el filtrado con los de "Violencia de género en todas sus modalidades distintas a la violencia familiar" no presenta ningun delito en ningun mes de ningun año para cada Municipio de cada Estado, siendo algo incoherente en el dataset, por lo cual no hace sentido hacer el análisis para este dataframe.

# **ÁNALISIS DE VIOLENCIA FAMILIAR EN CIUDAD DE MÉXICO**

Actualmente no conocemos las herramientas necesarias para poder realizar un análisis profundo a cada uno de los Estados que conforman a la República Mexicana, dado que para ello es necesario tener una representación gráfica de los resultados en vez de verlos directamente desde los dataframes. Para esto seleccionamos la Ciudad de México para enfocarnos en tres diferentes puntos, las cuales son presentadas en breve.

In [53]:
df_violencia_familiar_ciudad_mexico = df_violencia_familiar[df_violencia_familiar["entidad"] == "Ciudad de México"]
df_violencia_familiar_ciudad_mexico = df_violencia_familiar_ciudad_mexico.reset_index(drop=True)
df_violencia_familiar_ciudad_mexico.head(3)

año  clave_entidad           entidad  ...  octubre noviembre diciembre
0  2015              9  Ciudad de México  ...       77        92        71
1  2015              9  Ciudad de México  ...       96        91        93
2  2015              9  Ciudad de México  ...       17        13        19

[3 rows x 21 columns]

## **Los Municipios con mayor número de delitos por año**


Para poder calcular los Municipios con mayor número de delitos por año, primero es necesario poder calcular los delitos de violencia familiar para cada Municipio anualmente, para esto como los casos estan divididos mensualmente, se usará el comando *iloc* en el cual se seleccionarán las columnas que hacen referencia a los meses para cada Municipio de Ciudad de México y realizaremos una suma en el eje horizontal y asignaremos este resultado a una nueva columna llamada "total_anual" que hace referencia a la suma de delitos de violencia familiar anualmente para cada Municipio en los últimos 5 años.

In [54]:
df_violencia_familiar_ciudad_mexico["total_anual"] = df_violencia_familiar_ciudad_mexico.iloc[:, 9:20].sum(axis=1)
df_violencia_familiar_ciudad_mexico.head(2)

año  clave_entidad           entidad  ...  noviembre diciembre total_anual
0  2015              9  Ciudad de México  ...         92        71         915
1  2015              9  Ciudad de México  ...         91        93         878

[2 rows x 22 columns]

Finalmente realizaremos una agregación usando el comando *groupby* en el cual agruparemos por Municipio y por Año en base a la suma de delitos anuales la cual llamamos "total_anual", y estos serán acomodados del valor mayor al menor (orden descendiente), dándonos así los Municipios con mayor número de delitos de violencia familiar por año.

In [55]:
mayor_violencia_familiar_anual = df_violencia_familiar_ciudad_mexico.groupby(['año', 'municipio']).agg({'total_anual':sum})
mayor_violencia_familiar_anual = mayor_violencia_familiar_anual['total_anual'].groupby('año', group_keys=False).nlargest(5)
mayor_violencia_familiar_anual

año   municipio        
2015  Iztapalapa           2909
      Gustavo A. Madero    1758
      Cuauhtémoc           1225
      Álvaro Obregón       1176
      Iztacalco            1078
2016  Iztapalapa           3665
      Gustavo A. Madero    1793
      Cuauhtémoc           1366
      Tlalpan              1285
      Álvaro Obregón       1212
2017  Iztapalapa           3404
      Gustavo A. Madero    1848
      Tlalpan              1385
      Cuauhtémoc           1285
      Álvaro Obregón       1110
2018  Iztapalapa           3082
      Gustavo A. Madero    1950
      Cuauhtémoc           1590
      Tlalpan              1477
      Álvaro Obregón       1420
2019  Iztapalapa           4269
      Gustavo A. Madero    2819
      Cuauhtémoc           1966
      Álvaro Obregón       1851
      Tlalpan              1740
2020  Iztapalapa           4346
      Gustavo A. Madero    3386
      Álvaro Obregón       1972
      Cuauhtémoc           1959
      Tlalpan              1710
Name: total_anua

Con este resultado, podemos concluir que Iztapalapa ha sido el Municipio de Ciudad de México que se ha mantenido con el mayor número de casos a lo largo de los últimos 5 años, y en que tiempos de COVID-19 esto se ha mantenido.

## **Los Municipios con menor número de delitos por año**

Para el caso de los Municipios con menor número de delitos por año, realizaremos la misma agrupación mediante el comando *groupby* en el cual agruparemos por *año* y *municipio* al *total_anual* que hace referencia al total de delitos de violencia familiar para cada Municipio por cada año. Sin embargo, estos serán ordenados de menor a mayor (orden ascendiente), para poder así tener los Municipios con menor número de delitos por cada uno de los últimos 5 años.

In [56]:
menor_violencia_familiar_anual = df_violencia_familiar_ciudad_mexico.groupby(['año', 'municipio']).agg({'total_anual':sum})
menor_violencia_familiar_anual = menor_violencia_familiar_anual['total_anual'].groupby('año', group_keys=False).nsmallest(5)
menor_violencia_familiar_anual

año   municipio             
2015  No Especificado            25
      Cuajimalpa de Morelos     205
      Milpa Alta                251
      Tláhuac                   448
      La Magdalena Contreras    465
2016  No Especificado            17
      Milpa Alta                215
      Cuajimalpa de Morelos     257
      La Magdalena Contreras    547
      Miguel Hidalgo            569
2017  No Especificado            12
      Milpa Alta                256
      Cuajimalpa de Morelos     260
      Miguel Hidalgo            537
      La Magdalena Contreras    633
2018  No Especificado            18
      Milpa Alta                326
      Cuajimalpa de Morelos     342
      Miguel Hidalgo            527
      La Magdalena Contreras    769
2019  No Especificado             0
      Milpa Alta                385
      Cuajimalpa de Morelos     495
      Miguel Hidalgo            829
      La Magdalena Contreras    835
2020  No Especificado             0
      Milpa Alta                421

En este caso podemos ver que dentro del dataframe se encuentran resultados los cuales tienen como Municipio no especificado, sin embargo seguido a este los Municipios de Milpa Alta y Cuajimalpa de Morelos se encuentran con el menor número de casos y manteniéndose ésto mismo en la pandemia actual.

## **Delitos totales anuales**

Finalmente, para poder calcular los delitos totales anuales simplemente haremos una agrupación del dataframe ya trabajado y filtrado de acuerdo a los delitos de violencia familiar de Ciudad de México por año únicamente. Esto nos brindará el *total_anual* que hace referencia a la suma de delitos de violencia familiar de todos los estados, agrupándolos por año.

In [57]:
delitos_violencia_familiar_total_anual = df_violencia_familiar_ciudad_mexico.groupby(['año']).agg({'total_anual':sum})
delitos_violencia_familiar_total_anual

total_anual
año              
2015        14722
2016        17045
2017        16937
2018        18418
2019        23601
2020        25280

Con éste último resultado podemos confirmar que la pandemia actual dió indicio a mayor número de casos de violencia familiar, siendo que por el resguardo de las familias dentro de la cuarentena, estas conviven mayor tiempo ocasionando mayor número de delitos de violencia familiar.
